acceptance/tests/facts/schema.rb

```diff
@@ -0,0 +1,88 @@
+test_name "Validate facter output conforms to schema" do
+  tag 'risk:high'
+
+  require 'yaml'
+  require 'ipaddr'
+
+  # @param
+  # @return
+  def validate_fact(schema_fact, schema_fact_value, output_fact, output_fact_value)
+    fact_type = output_fact_value.class.to_s
+    schema_fact_type = schema_fact ? schema_fact_value["type"].capitalize : nil
+    if fact_type == "Hash"
+      fact_type = "Map"
+    end
+    if fact_type == "TrueClass" || fact_type == "FalseClass" #boolean.class returns FalseClass or TrueClass
+      fact_type = "Boolean"
+    end
+    if fact_type == "Float"
+      fact_type = "Double"
+    end
+    ipv4_regex = Regexp.new('^((25[0-5]|(2[0-4]|1\d|[1-9]|)\d)\.?\b){4}$')
+    ipv6_regex = Regexp.new('^(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))$')
+    mac_regex = Regexp.new('^([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})$')
+    if schema_fact_type == "Ip"
+      fact_type = ipv4_regex.match?(output_fact_value) ? "Ip" : fact_type
+    end
+    if schema_fact_type == "Ip6"
+      fact_type = ipv6_regex.match?(output_fact_value) ? "Ip6" : fact_type
+    end
+    if schema_fact_type == "Mac"
+      fact_type = mac_regex.match?(output_fact_value) ? "Mac" : fact_type
+    end
+    if fact_type == "Map"
+      schema_elements = schema_fact_value["elements"]
+      output_fact_value.each do |fact, value|
+        if value.nil? || !schema_elements
+          puts "skipped"
+          puts "fact: #{fact} fact value: #{value}"
+          puts "no schema elements: #{schema_elements} \n\n"
+          next
+        end
+        schema_fact, schema_fact_value = get_fact(schema_elements, fact)
+        validate_fact(schema_fact, schema_fact_value, fact, value)
+      end
+    end
+    if fact_type != schema_fact_type
+      puts "#{output_fact} has value: #{output_fact_value} and type: #{fact_type} does not conform to schema fact value type: #{schema_fact_type} \n\n"
+    end
```

Comment:
Thinking we should also validate that it's possible to parse `output_fact_value` based on the `fact_type`. For example, `identity.uid` is supposed to be an `integer`, so I would expect a fact value of `uid => "bob"` to fail validation. Doing `Integer(output_fact_value)` will either successfully parse the value as an integer or raise an exception.

For Hash/Map, I think you want to verify that `output_fact_value.is_a?(Hash)` before calling `output_fact_value.each do |fact, value|`.

For IP addresses, you may want to use ruby's built in classes like:

```ruby
if schema_fact_type == "Ip"
  fail_test("...") unless IPAddr.new(output_fact_value).ipv4?
elsif schema_fact_type == "Ip6"
  fail_test("...") unless IPAddr.new(output_fact_value).ipv6?
...
```

Reply:
Sorry, I think I got a little confused here. If facter were to emit `uid => "bob"` then the String "bob" would be passed as the `output_fact_value`. We'd set the `fact_type = 'string'`, but that won't match the type as specified in the schema. So it would be caught by the `assert_match(fact_type, schema_fact_type, ...`